## Introduction

This is the main repository containing all the scripts for the project: MultiOmics-ExomeSeq-Phenotype. Here, I am developing a Multi-Omics method that scores genomic variants responsible for a particular phenotype in the patient (**Note: The project is still in progress, and more scripts will be added, such as those that provide scoring component for the machine learning step**). At a later stage, I will try to combine all the individual scripts into a single pipeline.
</br>

- [Example Usage](#example-usage-and-details-of-the-scripts)
   - [UniProt Parser](#uniprotparser)
   - [Protein-Protein Interaction Parser](#ppiparser)
   - [PPI Experiment Count](#ppiexpcount)
   - [Interactome generator](#interactome)
   - [Module Input File Generator](#modulefile)
   - [Uniprot2ENSG Mapper](#uniprotensgmapper)
   - [Machine Learning ScoreComponent-1](#MLScoreComp1)
- [Arguments](#Arguments)
- [Output Details](#Output-Details)
- [Metadata files](#metadata-files)
- [Dependencies](#dependencies)
- [License](#license)

## Example Usage and Details of the Scripts

<a name="uniprotparser"></a>**UniProt Parser**

```console
python 1_uniprot_parser.py --inUniprot uniprot_sprot.dat --outPrimAC uniprot_main.tsv --outSecAC uniprot_secondary.tsv --outGeneID geneID.tsv
```    
</br>

- This will basically parse a UniProt file (Ex: uniprot_sprot.dat)
- Extracts the required data from each record
- Populates the output files with the processed data.

</br>

<a name="ppiparser"></a>**Protein-Protein Interaction Parser**

```console
python 2_interaction_parser.py --inInteraction intact.txt --inPrimAC uniprot_main.tsv --inSecAC uniprot_secondary.tsv --inGeneID geneID.tsv
```  
</br>

- Parses a miTAB 2.5 or 2.7 file (Ex: intact.txt)
- Maps the data of each protein-protein interaction experiment to the output files produced by `1_uniprot_parser.py`
- Extracts the UniProt accessions of the interacting proteins as well as other associated data
- Prints to STDOUT in tab-separated format

</br>

<a name="ppiexpcount"></a>**PPI Experiment Count**

```console
python 3_check_HumanPPIExp.py < intact.txt
```                      

</br>

- Parses a miTAB 2.5 or 2.7 file (Ex: intact.txt)
- Prints the count of Human-Human Protein Interaction experiments to STDOUT

</br>

<a name="interactome"></a>**Interactome generator**

```console
python 4_Interactome.py --inCuratedFile curatedPPI_BioGRID.tsv curatedPPI_Intact.tsv --inPrimAC uniprot_main.tsv --inCanonicalFile canonicalTranscripts_220221.tsv
```                      

</br>

- Parses the UniProt Primary Accession file (Ex: uniprot_main.tsv), canonical transcripts file (Ex: canonicalTranscripts_220221.tsv) & the output files (Ex: curatedPPI_BioGRID.tsv and curatedPPI_Intact.tsv) produced by `2_interaction_parser.py`
- Builds a high-quality human interactome
- Further, maps the UniProt accessions to ENSG using the `canonical transcripts file` and prints to STDOUT
- To produce the `canonical transcripts file`, please refer to [grexome-TIMC-Secondary](https://github.com/ntm/grexome-TIMC-Secondary/tree/master/Transcripts_Data)
</br></br>
**Note:** We consider only true binary interactions, and eliminate most of the complex expansion data while building the Interactome.

</br>

<a name="modulefile"></a>**Module Input File Generator**

```console
python 5_ModuleInputFile.py < Interactome_human.tsv
```                      

</br>

- Parses the output (Ex: Interactome_human.tsv) produced by `4_Interactome.py`
- Assigns a weight to each interaction and prints to STDOUT in .tsv format.
- This can be used as INPUT for most of the module identification/clustering methods

</br>

<a name="uniprotensgmapper"></a>**UniProt2ENSG Mapper**

```console
python 6_Uniprot2ENSG.py --inPrimAC uniprot_main.tsv --inCanonicalFile canonicalTranscripts_220221.tsv
```                      

</br>

- Parses the output files produced by `1_uniprot_parser.py` and the `canonical transcripts file` (Ex: canonicalTranscripts_220221.tsv)
- Maps UniProt accession to ENSG and prints to STDOUT

</br>

<a name="MLScoreComp1"></a>**Machine Learning ScoreComponent-1**

```console
python 7_Score1ML_candidateGenes.py --inPrimAC uniprot_main.tsv --inCandidateFile candidateGenes.xlsx --inCanonicalFile canonicalTranscripts_220221.tsv --inInteractome Interactome_human.tsv
```                      

</br>

- Parses the UniProt Primary Accession file (Ex:uniprot_main.tsv), canonical transcripts file (Ex: canonicalTranscripts_220221.tsv), Candidate Genes file (Ex: candidateGenes.xlsx) & Interactome file generated by `Interactome.py`
- Checks the number of Interactors for each gene
- Checks the number of Interactors that are known candidate genes and calculates P-values
- Further, computes Benjamini-Hochberg adjusted P-value and prints to STDOUT in .tsv format
- This script provides the first scoring component for the Machine Learning Step

## Arguments
```shell
--inUniprot                          UniProt Input File name
--outPrimAC                          UniProt Primary Accession Output File name
--outSecAC                           UniProt Secondary Accession Output File name
--outGeneID                          UniProt GeneID Output File name
--inInteraction                      miTAB 2.5 or 2.7 Input File name (Protein-Protein Interaction File)
--inCuratedFile                      Curated Output File name(s) (produced by 2_interaction_parser.py)
--inCanonicalFile                    Canonical Transcripts Input File name
--inCandidateFile                    Candidate Gene Input Files(s) name (.xlsx)
--inInteractome                      High-Quality Interactome Input File name (produced by 4_Interactome.py)
-h, --help                           Show the help message and exit
```

## Output Details

- For detailed description on the output generated, please use the help option with the scripts - `python script.py --help` OR `python script.py -h`


## Metadata files

- Currently, the scripts use only one metadata file i.e. `candidateGenes.xlsx`. Later, description of other files will be added.

1. candidateGenes.xlsx: </br>
  * Lists known candidate genes. This eases the identification of a patient's likely causal variant: variants impacting a known candidate gene can be easily selected.  
  * Required columns: </br>
    - Gene: name of gene (should be the HGNC name, see www.genenames.org).
    - pathologyID: pathology/phenotype
    - Confidence score: indicates how confident you are that LOF variants in this gene are causal for this pathology. Value: integers from 1 and 5 (5 meaning the gene is definitely causal, while 1 is a lower-confidence candidate).


## Dependencies

* Requires **Python version â‰¥ 3**
* External dependencies are kept to minimum in all the scripts. The only required python modules are listed below: </br>
  - Pandas ([Installation Guide](https://pandas.pydata.org/docs/getting_started/install.html))
  - SciPy ([Installation Guide](https://scipy.org/install/))
* Most other standard core modules should already be available on your system

## License

Licensed under GNU General Public License v3.0 (Refer LICENSE file for more details)
